go/vumitools/contact/models.py

```diff
@@ -240,42 +240,55 @@ def get_group(self, key):
         return self.groups.load(key)
 
     @Manager.calls_manager
-    def get_contacts_for_group(self, group):
-        """Return contact keys for this group."""
-        contacts = set([])
-        static_contacts = yield self.get_static_contacts_for_group(group)
-        contacts.update(static_contacts)
-        if group.is_smart_group():
-            dynamic_contacts = yield self.get_dynamic_contacts_for_group(group)
-            contacts.update(dynamic_contacts)
-        returnValue(list(contacts))
-
-    @Manager.calls_manager
     def get_contacts_for_conversation(self, conversation):
         """
         Collect all contacts relating to a conversation from static &
         dynamic groups.
         """
+        # TODO: FIXME: Avoid building up the whole set in memory.
         # Grab all contacts we can find
         contacts = set([])
         for groups in conversation.groups.load_all_bunches():
             for group in (yield groups):
-                group_contacts = yield self.get_contacts_for_group(group)
-                contacts.update(group_contacts)
+                index_page = yield self.get_contact_keys_for_group(group)
+                while index_page is not None:
+                    contacts.update(list(index_page))
+                    index_page = yield index_page.next_page()
 
         returnValue(list(contacts))
 
-    def get_static_contacts_for_group(self, group):
+    @Manager.calls_manager
+    def get_contact_keys_for_group(self, group):
+        """Return contact keys for this group."""
+        index_page = yield self.get_static_contact_keys_for_group(group)
+        if group.is_smart_group():
+            search_page = yield self.get_dynamic_contact_keys_for_group(group)
+            returnValue(ChainedIndexPages(
+                self.manager, search_page, index_page))
+        else:
+            returnValue(index_page)
+
+    def get_static_contact_keys_for_group(self, group):
         """
         Look up contacts through Riak 2i
         """
-        return group.backlinks.contacts()
+        return group.backlinks.contact_keys()
 
-    def get_dynamic_contacts_for_group(self, group):
+    def get_dynamic_contact_keys_for_group(self, group):
         """
         Use Riak search to find matching contacts.
         """
-        return self.contacts.raw_search(group.query).get_keys()
+        return self.search_contacts(group.query)
+
+    def search_contacts(self, query):
```

Comment:
The other method names seem to be explicit about returning contact keys instead of contact objects. Maybe this method name should too? `search_contacts` could imply the objects are returned, I think.